### PR TITLE
TRD-1708: Clean up Incoming hotpath

### PR DESCRIPTION
**Pull request by AlexandrosKyriakakis**

### CODE REVIEW

File: in_session.go [hunk 1]
```diff
@@ -27,6 +27,16 @@ type inSession struct{ loggedOn }
...
+
+	if session.CleanIncomingHotPath && IsExecutionReport(msg.rawMessage.Bytes()) {
+		if err := session.application.FromApp(msg, session.sessionID); err != nil {
+			return handleStateError(session, err)
+		}
+		if err := session.store.IncrNextTargetMsgSeqNum(); err != nil {
+			return handleStateError(session, err)
+		}
+		return state
+	}
```
Comment (sai-g) [inline in_session.go]: I don't think we need this check here; in the below switch, it's already checking for the message type. The main changes can be at line 64 below

File: session_state.go [hunk 2]
```diff
@@ -75,20 +76,39 @@ func (sm *stateMachine) Disconnected(session *session) {
...
+	}
 
-	session.log.OnIncoming(m.bytes.Bytes())
+	session.log.OnIncoming(rawBytes)
 
 	msg := NewMessage()
-	if err := ParseMessageWithDataDictionary(msg, m.bytes, session.transportDataDictionary, session.appDataDictionary); err != nil {
-		session.log.OnEventf("Msg Parse Error: %v, %q", err.Error(), m.bytes)
-	} else {
+	if session.CleanIncomingHotPath && IsExecutionReport(rawBytes) {
```
Comment (sai-g) [inline session_state.go]: Could you add a benchmark and see how much improvement we receive over the ParseMessageWithDataDictionary function?

File: in_session.go (hunk 1 above)
Comment (AlexandrosKyriakakis) [reply in_session.go]: On line 40 we do https://github.com/alpacahq/quickfix/blob/75f208b87f5f26139db3883716f399ee31c888cb/in_session.go#L40-L40 which is using 
https://github.com/alpacahq/quickfix/blob/75f208b87f5f26139db3883716f399ee31c888cb/field_map.go#L123-L123
But tagLookup is being initialised inside `ParseMessageWithDataDictionary`
https://github.com/alpacahq/quickfix/blob/75f208b87f5f26139db3883716f399ee31c888cb/message.go#L227-L228

So I would avoid removing this check. What do you think?

File: session_state.go (hunk 2 above)
Comment (AlexandrosKyriakakis) [reply session_state.go]: Added a benchmark for the whole `session.Incoming()`
Addressed here: https://github.com/alpacahq/quickfix/pull/17/commits/c835cd472ea2f2bf51094bfab76420905d79d0be

Comment (AlexandrosKyriakakis) [reply session_state.go]: Try changing the `CleanIncomingHotPath` variable true/false

Comment (AlexandrosKyriakakis) [reply session_state.go]: CleanIncomingHotPath = true
Where most of the time is spent in `mockApp.FromApp` (more than 10K ns)
```
/private/var/folders/cp/r9_v98l942x0jxt4d4fqyry80000gn/T/GoLand/___2BenchmarkIncoming_in_github_com_quickfixgo_quickfix.test -test.v -test.paniconexit0 -test.bench ^\QBenchmarkIncoming\E$ -test.run ^$
goos: darwin
goarch: arm64
pkg: github.com/quickfixgo/quickfix
BenchmarkIncoming
BenchmarkIncoming-10    	   83115	     13667 ns/op	    7983 B/op	      78 allocs/op
PASS

```

CleanIncomingHotPath = false
```
/private/var/folders/cp/r9_v98l942x0jxt4d4fqyry80000gn/T/GoLand/___2BenchmarkIncoming_in_github_com_quickfixgo_quickfix.test -test.v -test.paniconexit0 -test.bench ^\QBenchmarkIncoming\E$ -test.run ^$
goos: darwin
goarch: arm64
pkg: github.com/quickfixgo/quickfix
BenchmarkIncoming
BenchmarkIncoming-10    	   14322	     84926 ns/op	   40275 B/op	     455 allocs/op
PASS
```

File: session_settings.go [hunk 3]
```diff
@@ -124,9 +124,9 @@ func (s SessionSettings) BoolSetting(setting string) (bool, error) {
 	}
 
 	switch stringVal {
-	case "Y", "y":
+	case "Y", "y", "true", "True":
```
Comment (neal) [inline session_settings.go]: I would avoid this as these configs are common between every quickfix client in other languages so better to keep consistency when possible

File: session_state.go [hunk 4]
```diff
@@ -75,20 +76,39 @@ func (sm *stateMachine) Disconnected(session *session) {
 	}
 }
 
+var msgTypeExecutionReport = []byte("\x0135=8\x01")
+
+func IsExecutionReport(m []byte) bool {
```
Comment (neal) [inline session_state.go]: perhaps these belong around here? https://github.com/alpacahq/quickfix/blob/8a2144a438445349de04c07d02fbb8bcb40df063/msg_type.go#L26

and avoid exporting unless required

File: session_state.go [hunk 5]
```diff
@@ -75,20 +76,39 @@ func (sm *stateMachine) Disconnected(session *session) {
...
+	return bytes.Contains(m, msgTypeExecutionReport)
```
Comment (neal) [inline session_state.go]: 35=8 should be the first field I believe but not 100% if it's a rule - if so, can also use `bytes.HasPrefix` instead

File: session_state.go [hunk 6]
```diff
@@ -75,20 +76,39 @@ func (sm *stateMachine) Disconnected(session *session) {
...
 func (sm *stateMachine) Incoming(session *session, m fixIn) {
 	sm.CheckSessionTime(session, time.Now())
 	if !sm.IsConnected() {
 		return
 	}
+	var (
+		rawBytesBuffer = m.bytes
+		rawBytes       []byte
+	)
+	if rawBytesBuffer != nil {
```
Comment (neal) [inline session_state.go:87]: when is this nil?

File: memorystore.go [hunk 7]
```diff
@@ -107,15 +107,22 @@ func (store *memoryStore) GetMessages(beginSeqNum, endSeqNum int) ([][]byte, err
...
+func (f *memoryStoreFactory) Create(sessionID SessionID) (MessageStore, error) {
 	m := new(memoryStore)
 	if err := m.Reset(); err != nil {
 		return m, errors.Wrap(err, "reset")
 	}
+	f.store = m
 	return m, nil
 }
 
+func (f *memoryStoreFactory) GetStore() MessageStore {
```
Comment (neal) [inline memorystore.go]: this file seems to have unrelated changes to this PR?

File: session_state.go (hunk 5 above)
Comment (neal) [reply session_state.go]: seems like it should be a rule but there is also the begin string and body length first: https://github.com/alpacahq/quickfix/blob/65ca3790cc5ce48ebdff2fae2b33d9689dd627cf/message.go#L187

File: in_session.go [hunk 8]
```diff
@@ -27,6 +27,16 @@ type inSession struct{ loggedOn }
 func (state inSession) String() string { return "In Session" }
 
 func (state inSession) FixMsgIn(session *session, msg *Message) sessionState {
+
+	if session.CleanIncomingHotPath && IsExecutionReport(msg.rawMessage.Bytes()) {
+		if err := session.application.FromApp(msg, session.sessionID); err != nil {
```
Comment (neal) [inline in_session.go:32]: so we are essentially skipping all verification checks being done here? https://github.com/alpacahq/quickfix/blob/3ce9479197a8f43fa301fbe43dc3864d4b761d43/session.go#L515

I understand the goal of this PR is to avoid parsing the message, but not sure about skipping these checks. Maybe we still populate the headers? to avoid too many allocs, you can try using sync.Pool for NewMessage()

Comment (neal) [reply in_session.go:32]: the existing benchmark for ParseMessage does not show allocs if a Message is already inited:

```
Running tool: /opt/homebrew/bin/go test -benchmem -run=^$ -bench ^BenchmarkParseMessage$ github.com/quickfixgo/quickfix

goos: darwin
goarch: arm64
pkg: github.com/quickfixgo/quickfix
BenchmarkParseMessage-10    	 1695877	       694.8 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/quickfixgo/quickfix	2.079s
```

File: memorystore.go (hunk 7 above)
Comment (AlexandrosKyriakakis) [reply memorystore.go]: Addressed: https://github.com/alpacahq/quickfix/pull/17/commits/b67179eface00c8add1755b81a973dd15c93cc49

File: session_settings.go (hunk 3 above)
Comment (AlexandrosKyriakakis) [reply session_settings.go]: Addressed: https://github.com/alpacahq/quickfix/pull/17/commits/5854a27f1a66c619174271d33d86ab11184c8767

File: session_state.go (hunk 4 above)
Comment (AlexandrosKyriakakis) [reply session_state.go]: Addressed: https://github.com/alpacahq/quickfix/pull/17/commits/bf235aa4acb080a65854a8511b2c603b25556815

File: in_session.go (hunk 8 above)
Comment (AlexandrosKyriakakis) [reply in_session.go:32]: The point on this PR is that we never expect an issue after log on. 
So if the message is a log on or admin or heartbeat, etc it goes through all the common checks. 
Only case we will have a problem, is if venue has any error and change sender sequence number unexpectedly while responding to an execution reports. Which has never happen last 5 years.


Also check the benchmarks here, the difference is huge (I removed `mockApp.FromApp`) -> 
With Parsing
```
GOROOT=/opt/homebrew/Cellar/go/1.21.5/libexec #gosetup
GOPATH=/Users/alexandros.kyriakakis/go #gosetup
/opt/homebrew/Cellar/go/1.21.5/libexec/bin/go test -c -o /private/var/folders/cp/r9_v98l942x0jxt4d4fqyry80000gn/T/GoLand/___BenchmarkIncoming_in_github_com_quickfixgo_quickfix.test github.com/quickfixgo/quickfix #gosetup
/private/var/folders/cp/r9_v98l942x0jxt4d4fqyry80000gn/T/GoLand/___BenchmarkIncoming_in_github_com_quickfixgo_quickfix.test -test.v -test.paniconexit0 -test.bench ^\QBenchmarkIncoming\E$ -test.run ^$
goos: darwin
goarch: arm64
pkg: github.com/quickfixgo/quickfix
BenchmarkIncoming
BenchmarkIncoming-10    	   14547	     83642 ns/op	   41112 B/op	     462 allocs/op
PASS
```

Without Parsing
```
/private/var/folders/cp/r9_v98l942x0jxt4d4fqyry80000gn/T/GoLand/___BenchmarkIncoming_in_github_com_quickfixgo_quickfix.test -test.v -test.paniconexit0 -test.bench ^\QBenchmarkIncoming\E$ -test.run ^$
goos: darwin
goarch: arm64
pkg: github.com/quickfixgo/quickfix
BenchmarkIncoming
BenchmarkIncoming-10    	 5196544	       224.1 ns/op	     488 B/op	       8 allocs/op
PASS
```

Comment (neal) [reply in_session.go:32]: I am not sure about the theory these allocs - have you benchmarked specific code paths or looked at the profile to confirm?

I tried to test your benchmark (without the PR changes) and I noticed it failing the msg verification checks which is why you see all the improvement by skipping it.

first it's throwing `CompID problem` so you need to fix the session creation to match the msg: `SessionID{BeginString: "FIX.4.2", TargetCompID: "ISLD", SenderCompID: "LPCA"}`

then it hits `SendingTime accuracy problem` so I just commented this line to skip it for testing: https://github.com/alpacahq/quickfix/blob/3ce9479197a8f43fa301fbe43dc3864d4b761d43/session.go#L635

then the benchmark looks much better:

```
goos: darwin
goarch: arm64
pkg: github.com/quickfixgo/quickfix
BenchmarkIncoming
BenchmarkIncoming-10    	  219832	      5600 ns/op	    9030 B/op	      30 allocs/op
PASS
ok  	github.com/quickfixgo/quickfix	1.784s
```

some of this can be improved if you use a sync.Pool for the Message to reuse the internal allocs.

so I don't think ParseMessage is really as bad as you think since re-using a pre-allocated Message improves this much more:

```
goos: darwin
goarch: arm64
pkg: github.com/quickfixgo/quickfix
BenchmarkIncoming
BenchmarkIncoming-10    	  463188	      2574 ns/op	     135 B/op	       3 allocs/op
PASS
ok  	github.com/quickfixgo/quickfix	2.327s
```

edit: actually this is still running into more seq num mismatch errors

Comment (neal) [reply in_session.go:32]: I updated the benchmark to resolve all rejects and the results are much better now: https://github.com/alpacahq/quickfix/compare/develop...bench/incoming-neal

```
❯ go test -benchmem -run=BenchmarkIncoming -bench=BenchmarkIncoming -v
goos: darwin
goarch: arm64
pkg: github.com/quickfixgo/quickfix
BenchmarkIncoming
BenchmarkIncoming-10    	  241856	      4902 ns/op	    8902 B/op	      27 allocs/op
PASS
ok  	github.com/quickfixgo/quickfix	2.554s
```

and using sync.Pool for the Message, allocs go down to zero:

```
goos: darwin
goarch: arm64
pkg: github.com/quickfixgo/quickfix
BenchmarkIncoming
BenchmarkIncoming-10    	  399391	      2998 ns/op	       7 B/op	       0 allocs/op
PASS
ok  	github.com/quickfixgo/quickfix	2.618s
```

File: session_state.go (hunk 6 above)
Comment (AlexandrosKyriakakis) [reply session_state.go:87]: I dont expect this to be nil

**Comment by AlexandrosKyriakakis** [comment]
closing for now
